Update and install components using v1 manifests

nrc opened the pull request:
WIP

nrc commented:
Closes https://github.com/pingcap-incubator/tiup/issues/202

july2993 commented on pkg/repository/v1_repository.go:
```diff
@@ -54,6 +65,87 @@ func NewV1Repo(mirror Mirror, opts Options, local v1manifest.LocalManifests) *V1
+}
+
+func (r *V1Repository) ensureManifests() error {
+	// TODO
```
this will ensure we can the fresh and right snapshot&index?

nrc replied:
It will ensure that we have an up to date index, root, and snapshot, i.e., that we can call updateComponentManifest

july2993 commented on pkg/repository/v1_repository.go:
```diff
@@ -54,6 +65,87 @@ func NewV1Repo(mirror Mirror, opts Options, local v1manifest.LocalManifests) *V1
+		return err
+	}
+	for _, spec := range specs {
+		manifest, err := r.updateComponentManifest(spec.ID)
```
`updateComponentManifest` assume we can get a higher version manifest from the mirror than the local one,
seems we can not call this directly

nrc replied:
Good catch!

nrc commented:
PTAL @july2993 The test  for `updateComponents` is still not done, but I think I won't finish it today and this PR is getting large and I don't everyone's branches diverging too much.

july2993 commented on pkg/repository/v1manifest/manifest.go:
```diff
@@ -258,18 +286,18 @@ func (manifest *Timestamp) SnapshotHash() FileHash {
-func (manifest *Snapshot) VersionedURL(url string) (string, error) {
+// VersionedURL looks up url in the snapshot and returns a modified url with the version prefix, and that file's length.
+func (manifest *Snapshot) VersionedURL(url string) (string, *FileVersion, error) {
 	entry, ok := manifest.Meta[url]
```
seems key of Meta if just a file name. like `index.json`

july2993 replied:
we should first extract the filename from the url parameter?

nrc replied:
I think all fields which identify a file in any manifest should be a URL, not just a filename. The reason is that if in the future we add directories, then the filename is not enough.

july2993 replied:
in https://github.com/pingcap-incubator/tiup/blob/master/doc/design/manifest.md#snapshotjson it's a filename,
also `tools/migrate` will generate as filename.
then we need to fix both

july2993 commented on pkg/repository/v1manifest/manifest.go:
```diff
@@ -198,7 +224,9 @@ func (s *SignedBase) isValid() error {
 	if s.Ty != ManifestTypeRoot {
-		if err := CheckExpire(s.Expires); err != nil {
+		if err := CheckExpiry(s.Expires); err != nil {
+			fmt.Println("foo5")
```
need to be removed finaly

july2993 commented on pkg/repository/v1manifest/manifest.go:
```diff
@@ -185,6 +191,26 @@ func (s *SignedBase) expiryTime() (time.Time, error) {
+	}
+	time2, err := time.Parse(time.RFC3339, m2.Base().Expires)
+	if err != nil {
+		fmt.Println("foo3")
```
ditto

july2993 commented on pkg/repository/v1manifest/manifest.go:
```diff
@@ -185,6 +191,26 @@ func (s *SignedBase) expiryTime() (time.Time, error) {
+func ExpiresAfter(m1, m2 ValidManifest) error {
+	time1, err := time.Parse(time.RFC3339, m1.Base().Expires)
+	if err != nil {
+		fmt.Println("foo2")
```
ditto
we can remove this king of log after as follow up of this PR?

july2993 commented on pkg/localdata/profile.go:
```diff
@@ -248,6 +248,20 @@ func (p *Profile) InstalledVersions(component string) ([]string, error) {
+			return true, nil
+		}
+	}
+	return false, err
```
`return false, nil`   will be more clear for me

july2993 commented on pkg/repository/v1_repository.go:
```diff
@@ -50,32 +60,134 @@ func NewV1Repo(mirror Mirror, opts Options, local v1manifest.LocalManifests) (*V
+			}
+		}
+
+		reader, err := r.downloadComponent(versionItem)
```
defer reader.Close()

nrc replied:
This is just a Reader, not a ReadCloser and doesn't need (and can't be) closed

july2993 commented on pkg/repository/v1_repository.go:
```diff
@@ -50,32 +60,134 @@ func NewV1Repo(mirror Mirror, opts Options, local v1manifest.LocalManifests) (*V
+	if err != nil {
+		return false, err
+	}
+	snapRootVersion := snapshot.Meta[v1manifest.ManifestURLRoot].Version
```
should use `ManifestFilenameRoot` ?

july2993 commented on pkg/repository/v1_repository.go:
```diff
@@ -50,32 +60,134 @@ func NewV1Repo(mirror Mirror, opts Options, local v1manifest.LocalManifests) (*V
+	if err != nil {
+		return false, err
+	}
+	snapIndexVersion := snapshot.Meta[v1manifest.ManifestURLIndex].Version
```
the same here, the key of Meta is the file name, so can not access by url

july2993 commented on pkg/repository/v1_repository.go:
```diff
@@ -121,194 +226,173 @@ func fnameWithVersion(fname string, version uint) string {
-// Precondition: the snapshot manifest exists and is up to date
+// Precondition: the snapshot and index manifests exist and are up to date.
+// Returns nil if the component manifest is up to date.
 func (r *V1Repository) updateComponentManifest(id string) (*v1manifest.Component, error) {
```
do you want to return the Component event the local one is up to date?
how can we get the v1manifest.Component?  will add some other kind of method?

and i think `UpdateComponents` still can not call this directly if it return `nil, nil` when the manifest is up to date

nrc replied:
The update functions leave the manifests on disk, so after calling `updateComponentManifest` and there is no error, then the guarantee is that the manifest on disk is the most recent (and validated). To access the manifest, use LocalManifests. You are correct, there is a bug in `UpdateManifests` such that we won't install a component if the manifest is up to date, I'll fix that...

nrc commented:
@july2993 thanks for the review! I've addressed most of the comments and rebased. I think that we disagree on whether a file key should be a filename or url. See https://github.com/pingcap-incubator/tiup/pull/230#discussion_r428552779

july2993 commented:
> @july2993 thanks for the review! I've addressed most of the comments and rebased. I think that we disagree on whether a file key should be a filename or url. See [#230 (comment)](https://github.com/pingcap-incubator/tiup/pull/230#discussion_r428552779)

in https://github.com/pingcap-incubator/tiup/blob/master/doc/design/manifest.md#snapshotjson it's a filename,
also tools/migrate will generate as a filename.
if using url then we need to fix both of this later.

see https://github.com/pingcap-incubator/tiup/pull/230#discussion_r428558605
I think should have some method like `GetComponentManifest` to make sure we can get the manifest(the internal make sure the returning one is update to date), the current `updateComponentManifest` looks should only be used when we get a new snapshot and update the componentManifest locally if need.
